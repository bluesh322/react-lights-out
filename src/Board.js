import React, { useState } from "react";
import Cell from "./Cell";
import "./Board.css";

const cloneDeep = require('lodash/cloneDeep');

/** Game board of Lights out.
 *
 * Properties:
 *
 * - nrows: number of rows of board
 * - ncols: number of cols of board
 * - chanceLightStartsOn: float, chance any cell is lit at start of game
 *
 * State:
 *
 * - board: array-of-arrays of true/false
 *
 *    For this board:
 *       .  .  .
 *       O  O  .     (where . is off, and O is on)
 *       .  .  .
 *
 *    This would be: [[f, f, f], [t, t, f], [f, f, f]]
 *
 *  This should render an HTML table of individual <Cell /> components.
 *
 *  This doesn't handle any clicks --- clicks are on individual cells
 *
 **/

function Board({ nrows, ncols, chanceLightStartsOn }) {
  const [board, setBoard] = useState(createBoard( nrows, ncols, chanceLightStartsOn));

  /** create a board nrows high/ncols wide, each cell randomly lit or unlit */
  function createBoard(nrows, ncols, chanceLightStartsOn) {
    let initialBoard = [];
    for(let i = 0; i < nrows; i++) {
      initialBoard.push(Array.from({length: ncols}).map(() => {
        if(Math.random() * 1 < chanceLightStartsOn) return true;
        else return false;
      }));
    }

    return initialBoard;
  }

  function hasWon() {
    return !board.some(x => x.includes(true));
  }

  function flipCellsAround(coord) {
    setBoard(oldBoard => {
      const [y, x] = coord.split("-").map(Number);

      const flipCell = (y, x, boardCopy) => {
        // if this coord is actually on board, flip it

        if (x >= 0 && x < ncols && y >= 0 && y < nrows) {
          boardCopy[y][x] = !boardCopy[y][x];
        }
      };

      let newBoard = cloneDeep(oldBoard);


      flipCell(y, x, newBoard);
      flipCell(y+1, x, newBoard);
      flipCell(y, x+1, newBoard);
      flipCell(y-1, x, newBoard);
      flipCell(y, x-1, newBoard);

      return newBoard;
    });
  }

  return hasWon() ? (
    <div>
      <h1>You Won!</h1>
    </div>
  ): (
    <table className="Board">
      <thead>
        <tr>
          <th>Lights Out Game</th>
        </tr>
      </thead>
      <tbody>
        {board.map((row, rowIdx) => 
        <tr>
          {row.map((cell, colIdx) => <Cell flipCellsAroundMe={flipCellsAround} isLit={cell} coords={`${rowIdx}-${colIdx}`}/>)}
        </tr>)}
      </tbody>
    </table>
  )
  
}

Board.defaultProps = {nrows:5,ncols:5,chanceLightStartsOn: .5}

export default Board;
