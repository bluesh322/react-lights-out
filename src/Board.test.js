import React from "react";
import { render, fireEvent } from "@testing-library/react";
import Board from "./Board";

it("smoke test Carousel component", () => {
  render(<Board/>);
});

it("snapshot test for Carousel component when all lights are lit", () => {
    const {asFragment} = render(<Board nrows={5} ncols={5} chanceLightStartsOn={1}/>);
    expect(asFragment()).toMatchSnapshot();
})

it("snapshot test for when none of the lights are lit --> You Win!", () => {
    const {asFragment} = render(<Board nrows={5} ncols={5} chanceLightStartsOn={0}/>);
    expect(asFragment()).toMatchSnapshot();
})

it("check if cell clicking is switching the correct lights", () => {
    const { getAllByRole} = render(<Board nrows={3} ncols={3} chanceLightStartsOn={1}/>);
    const cells = getAllByRole("button");

    // all cells start out as lit
    cells.forEach(cell => {
    expect(cell).toHaveClass("Cell-lit");
    });

    // click on the middle cell
    fireEvent.click(cells[4]);
    
    // now only cells at the corners should be lit
    let litIndices = [0, 2, 6, 8];
    cells.forEach((cell, idx) => {
    if (litIndices.includes(idx)) {
        expect(cell).toHaveClass("Cell-lit");
    } else {
        expect(cell).not.toHaveClass("Cell-lit");
    }
    });
});