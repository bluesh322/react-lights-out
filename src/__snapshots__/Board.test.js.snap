// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test for Carousel component when all lights are lit 1`] = `
<DocumentFragment>
  <table
    class="Board"
  >
    <thead>
      <tr>
        <th>
          Lights Out Game
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
      </tr>
      <tr>
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
        <td
          class="Cell Cell-lit"
          role="button"
        />
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;

exports[`snapshot test for when none of the lights are lit --> You Win! 1`] = `
<DocumentFragment>
  <div>
    <h1>
      You Won!
    </h1>
  </div>
</DocumentFragment>
`;
